pennylane/devices/capabilities.py

```diff
@@ -0,0 +1,328 @@
+            controllable="controllable" in properties,
+            differentiable="differentiable" in properties,
+            conditions=conditions,
+        )
+    return operators
+
+
+def _get_operations(document: dict, prefix: str = "") -> dict[str, OperatorProperties]:
+    """Gets the supported operations from a TOML document."""
+    section = _get_toml_section(document, "operators.gates", prefix)
+    return _get_operators(section)
+
+
+def _get_observables(document: dict, prefix: str = "") -> dict[str, OperatorProperties]:
+    """Gets the supported observables from a TOML document."""
+    section = _get_toml_section(document, "operators.observables", prefix)
+    return _get_operators(section)
+
+
+def _get_measurement_processes(
+    document: dict, prefix: str = ""
+) -> dict[str, list[ExecutionCondition]]:
+    """Gets the supported measurement processes from a TOML document."""
+
+    section = _get_toml_section(document, "measurement_processes", prefix)
+    measurement_processes = {}
+    iterator = section.items() if hasattr(section, "items") else zip(section, repeat({}))
+    for mp, attributes in iterator:
+        if unknowns := set(attributes) - {"conditions"}:
+            raise InvalidCapabilitiesError(
+                f"Measurement '{mp}' has unknown attributes: {list(unknowns)}"
+            )
+        condition_strs = attributes.get("conditions", [])
+        if unknowns := set(condition_strs) - VALID_CONDITION_STRINGS:
+            raise InvalidCapabilitiesError(
+                f"Measurement '{mp}' has unknown conditions: {list(unknowns)}"
+            )
+        conditions = [ExecutionCondition(c) for c in condition_strs]
+        _validate_conditions(conditions)
+        measurement_processes[mp] = conditions
+    return measurement_processes
+
+
+def _get_compilation_flags(
+    document: dict, prefix: str = "", default: bool = True
+) -> dict[str, bool]:
+    """Gets the boolean capabilities in the compilation section."""
+
+    section = _get_toml_section(document, "compilation", prefix)
+
+    if unknowns := set(section) - VALID_COMPILATION_FLAGS.keys():
+        raise InvalidCapabilitiesError(
+            f"The compilation section has unknown options: {list(unknowns)}"
+        )
+
+    if not section.get("overlapping_observables", True) and section.get(
+        "non_commuting_observables", False
+    ):
+        raise InvalidCapabilitiesError(
+            "When overlapping_observables is False, non_commuting_observables cannot be True."
+        )
+
+    if not default:
+        return dict(section)
+
+    return {flag: section.get(flag, default) for flag, default in VALID_COMPILATION_FLAGS.items()}
+
+
+def _get_options(document: dict) -> dict[str, str]:
+    """Get custom options"""
+    return document.get("options", {})
+
+
+def parse_toml_document(document: dict) -> DeviceCapabilities:
+    """Parses a TOML document into a DeviceCapabilities object."""
```

Comment:
What this returns will always exclude anything that needs to be accessed with a `prefix`, right? And in that case, could we clarify that in the documentation here?